### ISSUE TITLE
Add protocol rule of `1 operation per DID per transaction`

### ISSUE THREAD

**Issue by thehenrytsai** [decentralized-identity/sidetree#173]

**Comment by OR13** [comment]
I have wondered about this.

This does seem like it could slow down a migration, where you want to anchor ops in the same order quickly.

**Comment by OR13** [comment]
I guess this issue is very related: https://github.com/decentralized-identity/sidetree-core/issues/191

Previously, I had been thinking that a given block might has 3 anchor transactions, and the same did might have create, update and revoke operations all  included in one block, but split across 3 ethereum transactions (3 anchorFiles and 3 batchFiles). Seems like we are saying we don't want that.

**Comment by thehenrytsai** [comment]
That's correct, this is a precautionary measure to prevent anyone spamming operations against a particular DID.

**Comment by OR13** [comment]
What do we do if we see multiple, no-op / warn on any after the first, or throw out the first as well?

**Comment by OR13** [comment]
To clarify, this appears to be only a thing that honest nodes can do (not build batches with multiple ops).

When pulling from the ledger, the presence of multiple ops, should probably just be handled normally, as this is a feature for cross ledger DID sync speed.

For example, I can build a single sidetree txn that only contains operations for my DID, and almost instantly migrate from Bitcoin to Ethereum, otherwise I have to wait blockTime for each op.

**Comment by thehenrytsai** [comment]
Hey @OR13, the goal for this issue was to really enforce 1 op per DID per transaction at the protocol level. If we see multiple ops, the thought was to penalize the writer and drop the entire batch.

This prevents anyone from writing millions of operations (invalid) against a particular honest DID.

I don't understand fully regarding the migration scenario you described, can you elaborate? @csuwildcat, can you also respond so we can move forward with this.

**Comment by OR13** [comment]
In order to migrate from one system to another, in theory, I need only replay my operations in the order the exist on the original ledger. The create will result in the same uniqueSuffix, subsequent ops will ensure that keys /  services are up to date. If you wanted to, you could deactivate the did on the older ledger as the final step and effectively migrate from say ethereum to bitcoin, and only the method name would change in your did.

In order to do this quickly, a user might want to run a full node, and create some batch files which contained the chain of operations to rapidly update the document without waiting block time. Its probably an edge case, and maybe the security benefit is worth loosing it.

I don't see a problem in enforcing the protocol rule of disregarding a batch which contains multiple operations for a did, but that would mean that signatures need to be validated on the way in.

Attacker can craft ops and submit them to honest nodes:

- multiple valid ops for one DID
- spoof ops for other DIDs, (signatures would not be correct).

Honest nodes will check op signatures as they receive them, and ensure that batches don't violate the protocol.

Attacker can still add batches directly which violate the protocol, but they will be ignored.

when processing a batch, only a little extra effort is required to ensure that there is only 1 op per batch, signatures are already processed at resolution time by default.

**Comment by csuwildcat** [comment]
@OR13 couple things to consider in relation to this:

1. The need to limit the ability to create a ton of spurious claims if op existence for a single DID is a must, so I wouldn't want to remove this safeguard.
2. You can't trust a migrated series of DID ops and stand on the fact that the meatspace owner of the ID on the new chain/system is the same as the original for the same reason you can't transfer ION DIDs in an effectively reliable way.

**Comment by OR13** [comment]
@csuwildcat 

1. I think we agree, anyone can craft and submit ops. limiting one op per DID per batchFile can be enforced by honest nodes, and batches that violate this, can be ignored, doing so would protect against a user just spamming tons of updates for their own DID but will require checking signatures when assembling the batchFile, this seems like a fine performance / security improvement.

2. I'm not sure I agree (or I'm missing something), you can't transfer a DID because each controller is their own root of trust, but that means that in theory they can create a new DID on a new ledger, and if they replay ops in the correct order they should end with a document in the same state, with a different method name, anchored to the new ledger, the same late publish attack would apply to both the original and the new ledger, so they could attack their own migration, but they can do the same to their original operation chain. 

If you don't trust signatures from keys rooted in the original ledger, you cannot trust signatures on the migrated ops, compromised DIDs can be migrated, but I think thats true of the original DID as well.

Maybe this hints at the value of late publish as a recovery mechanism of last resort.

**Comment by thehenrytsai** [comment]
> Maybe this hints at the value of late publish as a recovery mechanism of last resort.
That's an interesting idea.

The point on migration is that yes, you can technically migrate IDs from one chain to another, but it would not be secure for Relying Party to switch trust of an identity from one chain to another without keeping knowledge on both.

**Comment by csuwildcat** [comment]
@OR13 Henry is correct in that you wouldn't be able to logically assume, as a Relying Party, that the controller of a migrated DID is the controller of that DID from the original system, unless you ran both systems and knew they weren't replaying an incomplete subset of ops.

The other thing you brought up is very interesting; the idea that a late publish could actually aid recovery. As I think about it, you may be onto something cool, but there's only one issue I see: you'd need to commit the value without disclosing it, so it pretty much would require you to do a single txn/op yourself, given you need assurance the batch isn't revealed to observing nodes before you want it to be.

**Comment by OR13** [comment]
I see the point about maintaining knowledge.

yes, currently this recover operation costs 0.00003 ETH on Ropsten. I implemented integration tests for the late publish scenario, finally feel like I understand why sidetree DIDs are not transferrable.

**Comment by cboscolo** [comment]
Regarding one DID operation per transaction.
Rather than using JSON array for storing the DIDs in the batch, can we use a map with DID suffix as the key, and `{}` as the value? (I referenced this in the optimization ticket I wrote. (https://github.com/decentralized-identity/sidetree/issues/237)

As a map, it's easy to check for the existence of one particular DID, with little loss of efficiency iterating over all keys, and no extra logic is needed to figure out if there are dupes. Thus, the only check needed is to ensure the number of entries in the map equals the number of operations in the batch.

If there is agreement, I can open a separate ticket for just this bit.
